### DISCUSSION TITLE
API design for special movement actions

### DISCUSSION THREAD

**Discussion by idanarye** [idanarye/bevy-tnua#24] [Brainstorm]
I want to figure out the design for things like double jumps, air dashes, wall jumps, vaulting, gliding, etc...

It feels like these should all be connected. A wall jump is like a double jump, but when a wall is nearby. And maybe with an horizontal boost? But an horizontal boost is a dash. And the mechanism for detecting when an air dash is allowed should be connected to the mechanism for detecting when a double jump is allowed.

So I want to come up with a good design that takes them all into account.

**Comment by idanarye** [comment]
# Initiating air actions

The instinctive way to design an API for a double jump is to add a `max_air_jumps: usize` field to `TnuaPlatformerConfig`. I find this design too restrictive:

1. When adding more actions, the configuration becomes too complex. Say I want to allow for one jump and one air dash - can a jump and a air dash happen both in the same jump? How do I represent this?
2. How would it support mechanics like air pickups that allow you to do an extra jump?
3. Should wall jumps reset the count?
4. Should all the air jumps be to the same height? This is usually not the case. So should the configuration hold an array?

Tnua aims to maximize customizability, and the way to do it is to move those decisions to user systems. This means that Tnua itself will _not_ know how many jumps the player is allowed to do midair. Instead, Tnua will provide a mechanism for the user system to determine the number of the next jump, and the user system will then select whether or not to jump and with which strength:
```rust
// Very pseudocode
controls.jump = if jump_button_pressed() {
    match next_midair_jump_number() {
        0 => Some(1.0), // jump from the ground
        1 => Some(0.6), // air jump is less powerful
        2 => Some(0.2), // Second air jump is even tinier
        _ => None // third air jump is not allowed
    }
} else {
    None
}
```

**Reply by idanarye** [reply]
I'm thinking about having two components:

* `TnuaAirActionsRegulator` - when this component is added, it will be in charge of deciding when the character can jump (or do other actions). This means that the character will not jump even when on the ground unless the user system specifically tells this component that it's okay to jump.
    This component will also contain information about when the character is grounded, so you can get the default behavior with it if you do something like:
    ```rust
    if air_actions_regulator.is_grounded() {
        air_actions_regulator.can_initiate_this_frame = true;
    }
    ```
* `TnuaAirActionsCounter` - this is not something Tnua will actually update (which means I might put it in `control_helpers`). I'll make it a component for ease of use in simple cases, but it could also be used inside user defined components:
    ```rust
    #[derive(Component, Default)]
    struct MyActionsCounter {
        jump: TnuaAirActionsCounter,
        dash: TnuaAirActionsCounter,
    }
    ```
    The user system will be responsible for updating it when needed:
    ```rust
    if air_actions_regulator.is_grounded() {
        my_actions_counter.jump.reset();
        my_actions_counter.dash.reset();
    }
    ```
    For the rest of this comment though, for the sake of simplicity, I'll work with a single counter (and only with jumps - no other air actions)

The idea, very vaguely, is to do something like this:

```rust
if jump_button_pressed() {
    let mut air_actions_counter = air_actions_counter.with(&mut air_actions_regulator);
    let jump = match air_actions_counter.current() {
        0 => Some(1.0),
        1 => Some(0.6),
        2 => Some(0.2),
        _ => None,
    };
    air_actions_counter.this_frame(jump.is_some());
    controls.jump = jump;
}
```

**Reply by idanarye** [reply]
I'm not actually 100% sure I _need_ `TnuaAirActionsRegulator`. I probably need something for the `TnuaAirActionsCounter`s to read from, but other than that why not _always_ allow jumping when `controls.jump.is_some()`?

Actually, I'll have to handle the case of repeated jumps when the player keeps holding the jump button. But maybe there are better ways to do that as well? For example, instead of `None` when the character is out of jumps we can send `Some(0.0)`?

**Reply by VincentDouchin** [reply]
I agree with the idea of a TnuaAirActionsCounter to know where we are for the air actions but I'm not sure we need .is_grounded() since we already have this information in the TnuaPlatformerAnimatingOutput with the jump_velocity. By default I think tnua should allow multiple jumps by default and only prevent them if specified to do so. The only thing I'm not sure about is if it makes senses to to use TnuaPlatformerAnimatingOutput to know wether the player is jumping or not, since in this case we're not using for animations.

**Reply by idanarye** [reply]
I don't think multiple jumps should be the default. It is extremely rare for a game to want to allow unlimited jumps, so virtually all games will need to regulate it - and if a game is making the extra effort to regulate it, it can easily put in the effort to configure the entity to allow air jumps. Even more so when that configuration consists of adding the very component that's used to regulate it.

Games that don't allow air jumps, on the other hand, are much more common. Even more common than games that allow double jumps - but that could be because it is simpler to implement and simpler to design. And since they require no extra regulation, I see no reason why not to make it the default behavior.

As for why I want `TnuaAirActionsRegulator.is_grounded` that's separate from `TnuaPlatformerAnimatingOutput` - I have several reasons:

1. Separation of concerns - "if you want to support air jumps you first need to add support for animation" is a very weird requirement.
2. `TnuaAirActionsRegulator` may already be needed in the control system, because it'll need it to okay the jump.
3. I'm thinking about preparing `TnuaAirActionsRegulator` in a different system, that runs before the user control system. `TnuaPlatformerAnimatingOutput` gets updated after the user control system, so the information there will be one frame late.

**Reply by VincentDouchin** [reply]
Yeah I see what you mean, even though I feel like it seems natural to check if the player is grounded before the jump it does make sense to only allow one by default. The only thing that bothers me is that the TnuaAirActionsRegulator would contain the same thing as the TnuaPlatformerAnimatingOutput, but is you plan to add more things to it for the other air actions it does make sense as well.

**Comment by idanarye** [comment]
# Types of air actions

Tnua already has jumps, and the air version should work the same as he ground action - the controls will specify the jump's height, and Tnua will calculate the initial vertical velocity required to reach that height.

How would the API for the other types of actions work?

Actions that I can think of:

* Dash - a quick horizontal movement for a short period of time.
    * Possibility to add a hop to the dash?
* Float - fall slower.
* Glide - similar to float, but with horizontal movement. The horizontal velocity is tied to the vertical one.
* Wall jump - a jump that also has some horizontal velocity.
* Wall cling
* Wall slide
* Wall run
* Vaulting
* Ledge hang
* Ledge shimmy
* Ledge climb
* "Zoom" (need better name?) - move the character quickly to a specific position and make it stay there. This is useful for certain platformer games that need to place the player exactly in a specific position (e.g. the top of a very narrow platform)

Note that many of these interact with the environment. My vision is that some other mechanism will tell the user control system how the environment looks like, and then it would create environment-unaware air actions.

These are a lot of actions, and I want to see if I can generalize them somehow.

**Reply by VincentDouchin** [reply]
Oh and I was thinking about a hover, like there is in Super Mario Land 2 Yoshi's Island, or like a jetpack, would something like this be a jump or implemented as a different action?

**Reply by idanarye** [reply]
Ideally I would have generalized **all** these actions into a single one, and have different Wiki entries that explain how to implement each one. I practice we would probable have to group them into several generalizations.

But I think _hover_ and _jetpack_ are similar in that regard. Both of them are about applying an upward force that's stronger than gravity, in order to keep the character above a certain vertical velocity. In hover's case that velocity is negative (or at least zero), and in jetpack it's positive (but still finite - we don't want to launch the character upward too fast), but that's just a number that the user control system will have to pass to Tnua.

Also, in both actions we may or may not want a strong initial boost to reach above a certain starting velocity. For example, in the Yoshi's island example, if it gets applied when the character is falling at full speed we want to quickly break the fall first and only then apply a smaller upward acceleration. And if a player applies a parachute ability while falling, we'll also want to break the fall quickly first before maintaining the vertical acceleration that would fit the rest of the parachuting.

One difference thing I've seen in Yoshi's Island is that once Yoshi reaches a certain positive upward velocity, the action ends. So maybe there needs to be some control on how long the ability is maintained after reaching the target vertical velocity?

Actually, maybe it needs to be chained? So hover will be (e.g., very pseudo):

```rust
[
    TnuaAction::ShiftToVelocity { vertical: -1.0, max_acceleration: 100.0 },
    TnuaAction::MaintainVelocity { vertical: -1.0 },
]
```

And Yoshi style hover:

```rust
[
    TnuaAction::ShiftToVelocity { vertical: -1.0, max_acceleration: 100.0 },
    TnuaAction::ShiftToVelocity { vertical: 1.0, max_acceleration: 5.0 },
]
```

Or something like that.

**Reply by idanarye** [reply]
I've managed to express most of them that way:
* Dash
  ```rust
  [
    TnuaAction::MaintainVelocity {
      horizontal: direction * 100.0,
      duration: 1.0, // seconds? units? need to decide
    },
    TnuaAction::ShiftToVelocity { horizontal: 0.0 },
  ]
  ```
* Float:
  * Parachute style
    ```rust
    [
      TnuaAction::ShiftToVelocity { vertical: -1.0 }, // block the fall
      TnuaAction::MaintainVelocity { vertical: -1.0 },
    ]
    ```
  * Yoshi style
    ```rust
    [
      TnuaAction::ShiftToVelocity { vertical: -1.0 }, // block the fall
      TnuaAction::ShiftToVelocity { vertical: 1.0, max_acceleration: 5.0 }, // climb section
    ]
    ```
* Wall jump
  ```rust
  [
    // should regular jumps be actions as well?
    TnuaAction::Jump {
      max_height: 4.0,
      set_horizontal_velocity: Vec3::new(-10.0, 0.0, 0.0),
    },
  ]
  ```
* Wall cling
  ```rust
  [
    TnuaAction::FixateTranslation {
      anchor: current_point_on_wall,
      fixation: TnuaFixation::Point,
    }
  ]
  ```
* Vaulting
  ```rust
  [
    TnuaAction::FixateTranslation {
      anchor: point_at_elevation, // decide each frame
      fixation: TnuaFixation::Plane { normal: Vec3::Y },
    },
  ]
  ```
* Ledge hang / shimmy
  ```rust
  [
    TnuaAction::FixateTranslation {
      anchor: point_on_ledge, // probably need to add offset to character center
      fixation: TnuaFixation::Axis(direction_the_ledge_is_going),
    },
  ]
  ```
* Ledge climb
  Option one:
  ```rust
  [
    TnuaAction::FixateTranslation {
      anchor: point_on_ledge + Vec3::Y * some_height + offset_from_ledge_to_where_character_hangs,
      fixation: TnuaFixation::Point,
      speed: 5.0,
    },
    TnuaAction::FixateTranslation {
      anchor: point_on_ledge + Vec3::Y * some_height,
      fixation: TnuaFixation::Point,
      speed: 5.0,
    },
  ]
  ```
  Alternatively:
  ```rust
  [
    TnuaAction::FollowSpline {
      path: some_cubic_curve,
      speed: 5.0,
    },
  ]
  ```
* Zoom
  ```rust
  [
    TnuaAction::FixateTranslation {
      anchor: point_to_zoom_to,
      fixation: TnuaFixation::Point,
    },
  ]
  ```

The ones I didn't

* Glide: this one can be more complex, and maybe it shouldn't be part of the regular platformer controller and have its own controller, with the platformer controller only used to detect when the character is grounded and should therefore stop gliding.
* Wall slide and wall run: these require two things to happen at once:
  1. `FixateTranslation` to snap the character to the wall.
  2. Limit the downward velocity.
  Do I create one complex action, or a compound action? I'm not sure it fits the action-sequence design very well...

I've also noticed that very few actions use more than a single action in the sequence, and the ones that do use it for entry/exit, so maybe we don't need to support a sequence (which does make things more complicated)?

**Reply by VincentDouchin** [reply]
I like it! For the jump it does make sense to have it as an action as well, instead of a special case. Why is the glide more complicated? Is it not a float parachute style where you maintain velocity horizontally?

**Reply by idanarye** [reply]
Gliding is not just parachuting with horizontal movement. It also means the vertical velocity is tied to the horizontal one - that you can dive down to get faster horizontally or climb up at the cost of horizontal speed.

**Comment by idanarye** [comment]
# Actions that use the environment

This is about actions like wall jumping, ledge hanging, vaulting over obstacles, etc.

My direction for design is that the actions themselves will **not** be aware of the environment geometry itself - Tnua will process a wall jump action the same even if there is no wall there. Instead, I'll provide a `TnuaSpatialSensor` that will tell the user control system about the surrounding obstacles.

It'd look something like that (as always - very pseudo. And the rest of the API may not work as depicted here):

```rust
if player_is_airborne() && player_is_pushing_jump() {
    if let Some(direction) = direction_player_is_pushing() {
        if let Some(surface) = spatial_sensor.surface_at_direction(direction) {
            control.action = Some(TnuaAction::WallJump {
                jump_height: 1.0,
                wall_position: surface.position,
                wall_normal: surface.normal,
            });
        }
    }
}
```

If a similar action was sent without a wall being there, Tnua would have made the character jump off an imaginary wall. It should not need the wall itself - only the data provided about it inside the action.

**Comment by idanarye** [comment]
# Multiple actions at once

What should happen if the user control system invokes multiple action motions at once? Should this be allowed? Won't the actions interfere with each other?

**Reply by VincentDouchin** [reply]
The way it seems to be done so far I don't think allowing multiple actions at the same time is the way to go. Is there even any game that allow this? But one thing to consider is how the inputs will be buffered, ie if I want to jump after a dash, how will this work?

**Reply by idanarye** [reply]
Some kind of cancellation mechanism (like in fighting games)? I think the user control system will query Tnua about the current action, and then instruct it about whether or not to cancel it.

At any rate, the reason I was thinking about combining actions was so wall jumps that add some horizontal velocity can be a combination of a jump and a dash. But maybe for these cases it would be simpler to have a single action that handles both.

**Comment by idanarye** [comment]
# Design idea - composition by axis

Maybe the actions will be more flexible if they can declare individual behavior on up to 3 axes?

A wallslide action, for example, will look like this:

```rust
actions.set([
    TnuaActionComponent {
        project_on: TnuaProjection::Plane { normal: Vec3::Y },
        operation: TnuaActionOperation::FixtateTranslation(position_where_the_character_slides),
        ..Default::default()
    },
    TnuaActionComponent {
        project_on: TnuaProjection::Axis { direction: Vec3::Y },
        operation: TnuaActionOperation::CapVelocity {
            direction: -Vec3::Y, // required despite the axis
            max_speed: 2.0, // this is the max slide speed. It's positive, but the direction is negative
        },
        max_acceleration: 5.0, // TBD - is this in addition to gravity?
        ..Default::default()
    },
]);
```

Any axes not covered by a `TnuaActionComponent` will just use Tnua's regular behavior.

**Comment by idanarye** [comment]
# The polymorphism solution

The action will be a trait object that accepts the action-less motor output (plus configuration, sensor-input, and custom state?) and can modify it.

This means that each action will have to be written manually, but I can just provide a library of helpers for all the common actions.

**Reply by idanarye** [reply]
Maybe the regular floating and the jumping can also be coded in the same way? That way they won't be calculated when an action is active. Also it'll probably simplify the code?

**Reply by idanarye** [reply]
```rust
pub trait TnuaAction {
    // Wish it could default to `()`, but apparently that's unstable
    type State;

    fn init_state(&self, ctx: &TnuaActionContext) -> Self::State;

    fn lifecycle_status(
        &self,
        state: &Self::State,
        ctx: &TnuaActionContext
    ) -> TnuaActionLifecycleStatus;

    fn apply(
        &self,
        state: &mut Self:::State,
        ctx: &mut TnuaActionContext,
        motor: &mut TnuaMotor,
    );
}
```

`TnuaActionContext` is the various data we'll pass to all the methods of the action:

```rust
pub struct TnuaActionContext {
    pub frame_duration: Duration,
    pub platformer_config: &TnuaPlatformerConfig,
    pub platformer_state: &mut TnuaPlatformerState, // actions may mutate this one
    // more data?
}
```

This can potentially grow in the future.

Each action has a `State`, which is a free form place where it can store whatever it wants. Tnua will create it with `init_state` when the action starts.

`apply` is the method that gets called each frame by `platformer_control_system` to actually do what the action is suppsed to do. It's job is to modify the motor (which was already set by Tnua's regular controller, but the action is free to completely override it), and of course it gets a mutable access to its own state.

Before `apply` is called though, `platformer_control_system` calls `lifecycle_status` which will return:

```rust
pub enum TnuaActionLifecycleStatus {
    /// The action cannot be stopped by anything the user does.
    Uninterruptible,
    /// The action will be stopped as soon as it no longer getting fed into the controls.
    Stoppable,
    /// The action can be cancelled into another action, but it'll continue if
    // the user simply doesn't provide any action.
    Cancellable,
    /// The action is finished, even if the user is still holding the button.
    // If the action is repeatable and the button is still held, a timer will
    // start counting until it can be repeated.
    Finished,
}
```

### Alternative API

```rust
pub trait TnuaAction {
    type State: Default;

    fn apply(
        &self,
        state: &mut Self:::State,
        ctx: &mut TnuaActionContext,
        being_replaced: TnuaActionLifecycleStatus,
        motor: &mut TnuaMotor,
    ) -> TnuaActionLifecycleDirective;
}
```

Here the action is getting a:
```rust
pub enum TnuaActionLifecycleStatus {
    /// There was no action in the previous frame
    Initiated,
    /// There was a different action in the previous frame
    CancelledFrom,
    /// This action was already active in the previous frame, and it keeps getting fed
    StillFed,
    /// This action was fed up until the previous frame, and now no action is fed
    NoLongerFed,
    /// This action was fed up until the previous frame, and now a different action tries to override it
    CancelledInto,
}
```

The action knows whether or not it is being replaced (though not by whom. I don't want to couple actions to each other), and can decide if it should keep going using:
```rust
pub enum TnuaActionLifecycleDirective {
    StillActive,
    Finished,
}
```

The other change here is that instead of having an `init_state` method, the state must have a default. The action will set the correct values for it on the first run, which is easily detectable by looking at `TnuaActionLifecycleStatus`.

This design means that when an action gets cancelled, both old and new actions will be able to modify the motor. But because they know about the existence of each other I'm going to leave this to the actions' discretion.

**Reply by idanarye** [reply]
`TnuaAction::State` will also be used for animating output, and in general I want to merge the animating output into `TnuaPlatformerState` (see #26)

**Comment by idanarye** [comment]
# ECS approach

Similar to the polymorphism approach, but uses ECS - modify the motor decision from user systems.

While I'm usually all for ECS solutions, in this case I don't like it:

* Adding and removing components when the user runs an action is not the most ECS-friendly thing to do - certainly not on an heavy archetype like the player character. Also, it'd require `apply_system_buffers` or else the control's responsiveness will be harmed. But maybe I'm just premature optimization?
* The Tuna-provided actions will have to use a system, but I'd like to keep the helpers opt-in - which would mean that the user will have to manually add that system.
* Will the opt-in use one `enum` component? Or will it use many different components?
* When user puts multiple magic components - what's going to happen? And this will happen when running one action before another has finished!
* The main benefit is the ability to freely query the world inside these systems - but my original design was about not needing to freely query the world (because that's already done the user control system)

I think some of these problems can be avoided by spawning a new entity for the action. But that's still more overhead than the polymorphic approach.

**Reply by VincentDouchin** [reply]
I'm not sure we need to have the components for querying, won't the current action already be accessible in the animation output? Like you I think that this is not the way to go.

**Reply by idanarye** [reply]
The querying I'm talking about is not the one done for animating (the animating is already done in a user system, and can query whatever it wants). It's the one done when calculating the actual physics for the action (what impulses and forces the motor should apply). An ECS approach will allow it to query whatever it needs, while a polymorphic[^1] approach only provides it with a predefined set of data, which I hope will be enough.

Of course, polymorphic actions will be able to bring some data in the action itself (this, e.g., how I'm planning wall actions to get data about the wall they are actioning against) - but that requires the user to pass that data in the control system, so they need to be conservative with what data they require.

[^1]: The other non-polymorphic approaches I've brought up are also only able to access a predefined set of data, but since they themselves are not user code it doesn't matter.

**Reply by VincentDouchin** [reply]
Oh alright, that makes more sense.

**Comment by SanyaMal4** [comment]
Thank you for your knowledge

**Reply by idanarye** [reply]
I don't feel I've shared much knowledge in this discussion. I'm mostly trying to figure out this topic myself here.

**Comment by idanarye** [comment]
# Concrete API suggestion (polymorphic actions, jump is just another action, unified with action counter)

```rust
controls.set_action(TnuaSetAction {
    identifier: "jump",
    action: tnua_actions::Jump(20.0), // move the height from the config to here, since it's no longer privileged?
    allowed: TnuaActionAllowed::Grounded {
        buffer: 0.2,
    },
});
```

This is a bit cumbersome, so maybe a fluent interface would be better:

```rust
// The default `allowed` is Grounded { buffer: 0.0 }
controls.set_action("jump", tnua_actions::Jump(20.0)).allow_buffer(0.2);
```

The identifier can be used to determine if the user switch action, to handle multiple action inputs (not combine them - just determine which one was the last to start), and to write in the animating output so that another system could handle the animation for the action. While these can be done with the `TypeId` of the action, I prefer a string because:

1. Distinct actions may have the same type with different parameters.
2. You can't match on `TypeId` when animating.

It'll work like this:

```rust
// Must be called before setting actions this frame. Or maybe just call it
// `begin_frame()` and have it reset the movement vector as well?
controls.action_frame_begin();

if dash_input() {
    controls.set_action("dash", tnua_actions::Dash(...));
}

if jump_input() {
    controls.set_action("jump", tnua_actions::Jump(20.0));
}
```

`"dash"` appears before `"jump"`, but that only matters if they happen to start in the same frame. Let's say the player jumped first, and while still holding the jump buttons they started a dash. So at first the jump action will be the only one set, which means it'll be the active one. Then, when the player hits dash, the dash action will override it. Tnua will remember that `"jump"` is still passed but was already overridden, so each frame it'll get passed again and just get ignored.

As for `TnuaActionAllowed` (or the fluent methods that replace it) - it'll have three options:

1. `Grounded` - the action may only start when the character is grounded. It may have a buffer, which means that the player may press the button that number of seconds before landing and the action will still take effect.
  * I'm considering whether coyote time should also be passed as part of `Grounded`?
2. `Always` - the action can always start, even if the player is airborne.
3. `Never` - the action cannot start. Why is this even useful? Because the action will still be registered in the currently registered actions, so if the button is still pressed and at a later point the action can start and `Grounded`/`Always` will be passed to it - Tnua will not initiate it. If the user does want it to be initiated in this case, they can just not set it instead of registering it with `TnuaActionAllowed::Never`.
  * Maybe `Never` should also have a buffer like `Grounded`?

Initially I thought about also having a `Regulated` which will help regulating the number of airborne actions, but now that I think about it it can just be done with `Grounded`/`Always`. Instead of a new component, I'm thinking of just exposing it from `TnuaPlatformerState`:

```rust
controls.begin_frame();

if tnua_state.is_grounded() {
    *jumps_counter = 0;
} else {
    match tnua_state.action_report() {
        TnuaActionReport::Started("jump") => {
            *jumps_counter += 1;
        },
        TnuaActionReport::Ongoing("jump") => {
            // Can also do stuff in this case, but not this time
        },
        _ => {}
    }

    // Player stepped from a cliff
    if jumps_counter == 0 {
        jumps_counter = 1;
    }
}

match jumps_counter {
    1 => {
        // Air jump is not as powerfull as regular jump
        controls.set_action("jump", tnua_actions::Jump(10.0)).allow_always();
    }
    2 => {
        // Second air jump is even less powerful
        controls.set_action("jump", tnua_actions::Jump(5.0)).allow_always();
    }
    _ => {
        // Note that this **is** the regular jump. If the player tries to jump
        // a third time in the air it won't jump and instead buffer a jump in
        // case they reach the ground soon enough after.
        controls.set_action("jump", tnua_actions::Jump(20.0)).allow_buffer(0.2);
    }
}
```

(note that `tnua_state.is_grounded()` is a good reason to leave the coyote time in `TnuaPlatformerConfig`. Though maybe it can return a duration since the last time the character was grounded?)

**Reply by idanarye** [reply]
Actually, scrap `begin_frame`. I'll just reset it in `platformer_control_system`.

**Reply by idanarye** [reply]
Maybe instead of having `TnuaSetAction` or a fluent interface, it'd be easier to have four methods?

```rust
controls.set_action_buffered("jump", tnua_actions::Jump(20.0), 0.2);
controls.set_action_grounded("jump", tnua_actions::Jump(20.0)); // proxy for a buffer of 0.0
controls.set_action_airborne("jump", tnua_actions::Jump(20.0));
controls.set_action_blocked("jump");
```

I hate the naming though, and can't think of better ones, so maybe I'll stay with `TnuaSetAction` after all. The only thing I like her is that unlike with `TnuaActionAllowed::Never`, with `set_action_blocked` I don't need to pass an actual action. So maybe keep just that one (instead of `TnuaActionAllowed::Never`)?

**Reply by idanarye** [reply]
BTW - I've used `tnua_actions::Jump(20.0)` in the examples instead of `tnua_actions::Jump(1.0)` and setting `full_jump_height: 20.0` in `TnuaPlatformerConfig`. I'm thinking of removing that field - see #25.

**Reply by idanarye** [reply]
If I want to make the regular movement itself an action too, I'll need another `TnuaActionAllowed` variant for it - let's call it `TnuaActionAllowed::Fallback`. Unlike the other types, `Fallback` will be usable even after being replaced by other actions - so the user control system can just _always_ pass it, and the the jump action finishes Tnua will fall back to the regular movement action.

I wonder if an override mechanism (where once an action starts it gets priority over actions that were already being fed) will work for the fallback actions. They'll need, of course, their own state - so that once a non-fallback action is over they can go back to it. But if that's not good enough, I can always add a `priority` field to resolve these things.

**Reply by idanarye** [reply]
Maybe instead of passing `TnuaActionAllowed` to `set_action`, I can add a method to the `TnuaAction` trait that will return it? After all, actions will usually have their own preferred value for it - `tnua_actions::Move` will always want to use `TnuaActionAllowed::Fallback` and `tnua_actions::DropThroughPlatform` will always want to use `TnuaActionAllowed::Grounded`. And this can easily be overwritten with a method:

```rust
controls.set_action("jump", tnua_actions::Jump(10.0).allow_always());
```
(note that this is a bit different than the similar looking line in the last snippet of the opening comment of this subdiscussion (https://github.com/idanarye/bevy-tnua/discussions/24#discussioncomment-6355381) because here `allow_always()` is a method of `tnua_actions::Jump` and there it was a method of a fluent API handle returned by `set_action`)

**Reply by idanarye** [reply]
`TnuaActionAllowed::Grounded` seems like a special case of something. Namely - it checks a condition (which happens to be "is the character grounded?") and then, based on that condition, acts as either `Always` or `Never`. Maybe the method should be the one checking that condition? In which case, the naming scene should probably change a bit...

**Reply by idanarye** [reply]
If regular movement becomes an action, the last thing to remain that needs the `TnuaPlatformerConfig` is rotation.

I don't want every action to have to consider the rotation, but I don't mind if just the base movement does. Maybe the fallback action can always happen (in which case its more of a "basis" action than a "fallback" one), and the other action (jump/dash/whatever) will just refrain from overriding the rotation (unless it's an action that does need to control rotation)?

**Reply by idanarye** [reply]
Handling multiple basis actions (actually, let's just call them "basis", and use "action" for the non-basis ones) may be trickier than I originally thought.
Say we have three basis actions - X, Y and Z.

* Frame 1: Only X is fed, so it is chosen.
* Frame 2: Both X and Y are fed, so Y is chosen because it is the most recent.
* Frame 3: X, Y and Z are fed, so Z is chosen because it is the most recent.
* Frame 4: Only X and Y are fed. What happens?

Obviously we want Y to be chosen. But to avoid storing all the fed actions in a big structure, we want to know that in advance. Say our control system looks like this:

```rust
controls.set_basis(X { ... });

if cond_for_y() {
    controls.set_basis(Y { ... });
}

if cond_for_z() {
    controls.set_basis(Z { ... });
}
```
When passing X, we don't know that Y is going to be passed in the same frame. With actions this is not an issue, because X was already done for and we'll just ignore it until we reach a frame where it does not happen. But now, we must store the `X { ... }` somewhere because we don't know if we'll need it or not.

Actually, the same holds for Y - we don't know yet if the Z will be called this frame, so it can't override it.

---

Because using multiple basis is already going to change the entire controls scheme (e.g. - you'd want different actions for a swim basis than for a regular movement basis) I think it is reasonable to make the user ensure they only pass one basis per frame. Maybe even panic if multiple basis are passed?
